ambry-frontend/src/main/java/com/github/ambry/frontend/AmbryIdConverterFactory.java

```diff
@@ -112,37 +113,97 @@ public Future<String> convert(RestRequest restRequest, String input, BlobInfo bl
         if (convertedId != null || exception != null) {
-          completeConversion(convertedId, exception, future, callback);
+          completeConverterOperation(convertedId, exception, future, callback, frontendMetrics.idConversionDownstreamCallbackTimeInMs);
         }
       }
       return future;
     }
 
+    @Override
+    public Future<Boolean> detectConflict(RestRequest restRequest, Callback<Boolean> callback) {
+      final CompletableFuture<Boolean> future = new CompletableFuture<>();
+      Exception exception = null;
+      frontendMetrics.idConverterDetectConflictRate.mark();
+      long startTimeInMs = System.currentTimeMillis();
+      try {
+        if (!isOpen) {
+          exception = new RestServiceException("IdConverter is closed", RestServiceErrorCode.ServiceUnavailable);
+        } else {
+          CallbackUtils.callCallbackAfter(detectConflict(restRequest),
+              (conflictStatus, e) -> completeConverterOperation(conflictStatus, e, future, callback, frontendMetrics.idConversionDetectConflictDownstreamCallbackTimeInMs));
+        }
+      } catch (Exception e) {
+        exception = e;
+      } finally {
+        frontendMetrics.deleteCallbackProcessingTimeInMs.update(System.currentTimeMillis() - startTimeInMs);
+        if (exception != null) {
+          completeConverterOperation(false, exception, future, callback, frontendMetrics.idConversionDetectConflictDownstreamCallbackTimeInMs);
+        }
+      }
+      return future;
+    }
+
+    /**
+     * Detects if the ID in the request can cause a conflict.
+     * @param restRequest @link RestRequest} representing the request.
+     * @return the {@link CompletionStage} that will be completed with the conflict status.
+     * @throws RestServiceException if the operation cannot be performed.
+     */
+    private CompletionStage<Boolean> detectConflict(RestRequest restRequest) throws RestServiceException {
+      final CompletionStage<Boolean> detectionFuture;
+      frontendMetrics.idConverterDetectConflictRate.mark();
+      if (restRequest.getRestMethod() == RestMethod.PUT && RestUtils.getRequestPath(restRequest)
+          .matchesOperation(Operations.NAMED_BLOB) && !RestUtils.isUpsertForNamedBlob(restRequest.getArgs())) {
+        // A named blob put without upsert is allowed only if the named blob does not exist.
+        // If this is a named blob put request without upsert, then we check if a mapping for the name already exists.
+        NamedBlobPath namedBlobPath = NamedBlobPath.parse(RestUtils.getRequestPath(restRequest), restRequest.getArgs());
+
+        detectionFuture = getNamedBlobDb().get(namedBlobPath.getAccountName(), namedBlobPath.getContainerName(),
+            namedBlobPath.getBlobName(), GetOption.Include_All).thenApply(namedBlobRecord -> {
+          LOGGER.info("A mapping for the namedBlob {} already exists. Original BlobId: {}",
```

Comment:
What about if the named blob is deleted?
If the named blob is deleted, we should allow upload even without the upsert option.
If it's the case, we need check the db record status.